src/emqx_batch.erl

```diff
@@ -0,0 +1,74 @@
+%% Copyright (c) 2018 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+
+-module(emqx_batch).
+
+-export([init/1, push/2, commit/1]).
+-export([size/1, items/1]).
+
+-type(options() :: #{
+        batch_size => non_neg_integer(),
+        linger_ms => pos_integer(),
+        commit_fun := function()
+       }).
+-export_type([options/0]).
+
+-record(batch, {
+          batch_size :: non_neg_integer(),
+          batch_q :: list(any()),
+          linger_ms :: pos_integer(),
+          linger_timer :: reference() | undefined,
```

Comment:
Maybe replace this with just a timestamp then expose an api to get age instead of starting timer in this module.
Side effects like this tend to lose control pretty easily and quickly.
E.g. when a push call is wrapped by a try catch, timer is started in try but will not be canceled in catch.